omniscidb/ResultSetRegistry/ResultSetRegistry.cpp

```diff
@@ -0,0 +1,267 @@
+  // because we will have all required buffers already allocated.
+  // TODO: To avoid having data cached here we don't need anymore, we should
+  // clean-up tokens we are not going to use in TemporaryTables of
+  // RelAlgExecutor.
+  if (useColumnarResults(*frag.rs)) {
+    if (frag.columnar_res) {
+      buf = frag.columnar_res->getColumnBuffers()[col_idx];
+    } else {
+      CHECK(frag.mutex);
+      std::lock_guard<std::mutex> columnaraize_lock(*frag.mutex);
+      if (!frag.columnar_res) {
+        std::vector<const hdk::ir::Type*> col_types;
+        for (size_t i = 0; i < frag.rs->colCount(); ++i) {
+          col_types.push_back(frag.rs->colType(i)->canonicalize());
+        }
+        frag.columnar_res =
+            std::make_unique<ColumnarResults>(frag.rs->getRowSetMemOwner(),
+                                              *frag.rs,
+                                              frag.rs->colCount(),
+                                              col_types,
+                                              0,
+                                              *config_);
+      }
+      buf = frag.columnar_res->getColumnBuffers()[col_idx];
+    }
+  } else if (frag.rs->isZeroCopyColumnarConversionPossible(col_idx)) {
+    buf = frag.rs->getColumnarBuffer(col_idx);
+  }
+
+  return buf ? std::make_unique<ResultSetDataToken>(frag.rs, buf, num_bytes) : nullptr;
+}
+
+bool ResultSetRegistry::useColumnarResults(const ResultSet& rs) const {
+  // We use ColumnarResults for all cases except those when can use
+  // zero-copy fetch or copyColumnIntoBuffer for all columns.
+  // It would be better to convert comlumns more lazily especially when
+  // columnar format is used and multiple columns might be fetched in
+  // parallel.
+  return !rs.isDirectColumnarConversionPossible() ||
+         rs.getQueryDescriptionType() != QueryDescriptionType::Projection ||
+         rs.areAnyColumnsLazyFetched();
+}
+
+TableFragmentsInfo ResultSetRegistry::getTableMetadata(int db_id, int table_id) const {
+  mapd_shared_lock<mapd_shared_mutex> data_lock(data_mutex_);
+  CHECK_EQ(db_id, db_id_);
+  CHECK_EQ(tables_.count(table_id), (size_t)1);
+  auto& table = *tables_.at(table_id);
+  mapd_shared_lock<mapd_shared_mutex> table_lock(table.mutex);
+  data_lock.unlock();
+
+  if (table.fragments.empty()) {
+    return getEmptyTableMetadata(table_id);
+  }
+
+  TableFragmentsInfo res;
+  res.setPhysicalNumTuples(table.row_count);
+  for (size_t frag_idx = 0; frag_idx < table.fragments.size(); ++frag_idx) {
+    auto& frag = table.fragments[frag_idx];
+    auto& frag_info = res.fragments.emplace_back();
+    frag_info.fragmentId = static_cast<int>(frag_idx + 1);
+    frag_info.physicalTableId = table_id;
+    frag_info.setPhysicalNumTuples(frag.row_count);
+    frag_info.deviceIds.push_back(0);  // Data_Namespace::DISK_LEVEL
+    frag_info.deviceIds.push_back(0);  // Data_Namespace::CPU_LEVEL
+    frag_info.deviceIds.push_back(0);  // Data_Namespace::GPU_LEVEL
```

Comment:
Looks like a bunch of unnecessary reallocations in a loop to me.

Reply:
Yes, `resize` is better here. Fixed.